File: akf_projects/akf_projects/page/project_dashboard/project_dashboard.js

// let filters = {};
// frappe.pages['project-dashboard'].on_page_load = async function(wrapper) {
// 	var page = frappe.ui.make_app_page({
// 		parent: wrapper,
// 		title: 'Project Dashboard',
// 		single_column: true
// 	});
// 	await filters.add(page);
// }


frappe.pages['project-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Project Dashboard',
        single_column: true
    });

    server_call.fetch_data(page);
};


server_call = {
    fetch_data: function(page) {
        frappe.call({
            method: "akf_projects.akf_projects.page.project_dashboard.project_dashboard.get_information",
            args: {},
            callback: function(r) {
				
                var info = r.message;
                console.log(info);

                // Render new content
                // $(frappe.render_template("project_survey_dashboard", {
                //     'region_wise_survey': info['region_wise_survey'],
                // })).appendTo(page.main);
				$(frappe.render_template("project_dashboard", r.message)).appendTo(page.main);

                _highcharts_.load_charts(info);
                // console.log(r.message);

            },
        });
    },
	
};



_highcharts_ = {
    load_charts: function (info) {      
      _highcharts_.region_wise_survey(info.region_wise_survey);     
      _highcharts_.district_wise_survey(info.district_wise_survey);     
      _highcharts_.tehsil_wise_survey(info.tehsil_wise_survey);     
      _highcharts_.product_wise_survey(info.product_wise_survey);     
      _highcharts_.project_status(info.project_status);     
      _highcharts_.approved_vs_unapproved_projects(info.approved_vs_unapproved_projects);     
    },

    region_wise_survey: function (data) {
        let chart_js = region_wise_survey_(data)
        $("#region_wise_survey").html(chart_js)
    }, 

    district_wise_survey: function (data) {
        let chart_js = district_wise_survey_(data)
        $("#district_wise_survey").html(chart_js)
    }, 

    tehsil_wise_survey: function (data) {
        let chart_js = tehsil_wise_survey_(data)
        $("#tehsil_wise_survey").html(chart_js)
    }, 

    product_wise_survey: function (data) {
        let chart_js = product_wise_survey_(data)
        $("#product_wise_survey").html(chart_js)
    }, 

    project_status: function (data) {
        let chart_js = project_status_(data)
        $("#project_status").html(chart_js)
    }, 

    approved_vs_unapproved_projects: function (data) {
        let chart_js = approved_vs_unapproved_projects_(data)
        $("#approved_vs_unapproved_projects").html(chart_js)
    }, 
     
};

function region_wise_survey_(data) {
	
    return `
    <script>

	Highcharts.chart('region_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'Region Wise Projects',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				allowPointSelect: true,
				cursor: 'pointer',
				depth: 35,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}

function district_wise_survey_(data) {
	
    return `
    <script>

	Highcharts.chart('district_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'District Wise Projects',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				allowPointSelect: true,
				cursor: 'pointer',
				depth: 35,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}


function tehsil_wise_survey_(data) {
	
    return `
    <script>

	Highcharts.chart('tehsil_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'Tehsil Wise Projects',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				innerSize: 100,
				depth: 45,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}

function product_wise_survey_(data) {
	
    return `
    <script>

	Highcharts.chart('product_wise_survey', {
		chart: {
			type: 'pie',
			options3d: {
				enabled: true,
				alpha: 45,
				beta: 0
			}
		},
		title: {
			text: 'Product Wise Projects',
			align: 'left'
		},
		subtitle: {
			text: null
		},
		plotOptions: {
			pie: {
				innerSize: 100,
				depth: 45,
				dataLabels: {
					enabled: true,
					format: '{point.name}: {y}'
				}
			}
		},
		series: [{
			type: 'pie',
			name: 'Region',
			data: ${JSON.stringify(data.data)},
		}],
		credits: {
			enabled: false 
		}
	});
	</script>
    `
}

function project_status_(data) {
	
    return `
    <script>

	Highcharts.chart('project_status', {
		chart: {
			type: 'column'
		},
		title: {
			text: 'Project Status'
		},
		subtitle: {
			text: null
		},
		xAxis: {
			categories: ['Project Status'],
			crosshair: true,
			accessibility: {
				description: 'Countries'
			}
		},
		yAxis: {
			min: 0,
			title: {
				text: null
			}
		},		
		plotOptions: {
			column: {
				pointPadding: 0.2,
				borderWidth: 0
			}
		},
		series: [
			{
				name: 'In Progress',
				data: [${JSON.stringify(data.in_progress)}]
			},
			{
				name: 'Completed',
				data: [${JSON.stringify(data.completed)}]
			},
			{
				name: 'Delayed',
				data: [${JSON.stringify(data.delayed)}]
			}
		]
	});
	</script>
    `
}

function approved_vs_unapproved_projects_(data) {
	
    return `
    <script>

	Highcharts.chart('approved_vs_unapproved_projects', {
		chart: {
			type: 'column'
		},
		title: {
			text: 'Approved VS Unapproved Projects'
		},
		subtitle: {
			text: null
		},
		xAxis: {
			categories: ['Project'],
			crosshair: true,
			accessibility: {
				description: 'Countries'
			}
		},
		yAxis: {
			min: 0,
			title: {
				text: null
			}
		},		
		plotOptions: {
			column: {
				pointPadding: 0.2,
				borderWidth: 0
			}
		},
		series: [
			{
				name: 'Pending',
				data: [${JSON.stringify(data.pending)}]
			},
			{
				name: 'Verified By Project Manager',
				data: [${JSON.stringify(data.approved_by_pm)}]
			},
			{
				name: 'Approved By CEO',
				data: [${JSON.stringify(data.approved_by_ceo)}]
			}
		]
	});
	</script>
    `
}

File: akf_projects/akf_projects/page/project_dashboard/project_dashboard.py
#main-dashboard
import frappe
from frappe import _


@frappe.whitelist(allow_guest=True)
def get_information(project=None):
    return {
        "region_wise_survey": get_region_wise_survey(project),
        "district_wise_survey": get_district_wise_survey(project),
        "tehsil_wise_survey": get_tehsil_wise_survey(project),
        "product_wise_survey": get_product_wise_survey(project),
        "survey_forms_count": get_total_survey_forms(project),
        "project_status": get_project_status(project),
        "approved_vs_unapproved_projects": get_approved_vs_unapproved_projects(project),
    }

def get_region_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN custom_region IS NULL OR custom_region = '' THEN 'N/A'
                ELSE custom_region
            END AS custom_region,
            COUNT(name) as survey_count
        FROM `tabProject`
        GROUP BY 
            CASE
                WHEN custom_region IS NULL OR custom_region = '' THEN 'N/A'
                ELSE custom_region
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("Region:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_district_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN custom_district IS NULL OR custom_district = '' THEN 'N/A'
                ELSE custom_district
            END AS custom_district,
            COUNT(name) as survey_count
        FROM `tabProject`
        GROUP BY 
            CASE
                WHEN custom_district IS NULL OR custom_district = '' THEN 'N/A'
                ELSE custom_district
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("district:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_tehsil_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN custom_tehsil IS NULL OR custom_tehsil = '' THEN 'N/A'
                ELSE custom_tehsil
            END AS custom_tehsil,
            COUNT(name) as survey_count
        FROM `tabProject`
        GROUP BY 
            CASE
                WHEN custom_tehsil IS NULL OR custom_tehsil = '' THEN 'N/A'
                ELSE custom_tehsil
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("tehsil:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_product_wise_survey(project=None):
    query = """
        SELECT 
            CASE
                WHEN custom_product IS NULL OR custom_product = '' THEN 'N/A'
                ELSE custom_product
            END AS custom_product,
            COUNT(name) as survey_count
        FROM `tabProject`
        GROUP BY 
            CASE
                WHEN custom_product IS NULL OR custom_product = '' THEN 'N/A'
                ELSE custom_product
            END      
    """
    result = frappe.db.sql(query, as_dict=False)
    # frappe.msgprint("product:")
    # frappe.msgprint(frappe.as_json(result))
    
    # total_surveys = sum(row[1] for row in result)
    total_surveys = 0
    return {"data": result, "total_surveys": total_surveys}

def get_project_status(project=None):
    # Query for projects with completed and in-progress status
    project_query = """
        SELECT 
            COUNT(CASE WHEN status = 'Completed' THEN 1 END) AS completed,
            COUNT(CASE WHEN status = 'Open' THEN 1 END) AS in_progress
        FROM `tabProject`
    """
    project_result = frappe.db.sql(project_query, as_dict=1)

    # Query to count projects that have any overdue tasks
    delayed_query = """
        SELECT COUNT(DISTINCT project) as delayed_project
        FROM `tabTask`
        WHERE status = 'Overdue'
        AND project IS NOT NULL
    """
    delayed_result = frappe.db.sql(delayed_query, as_dict=1)
    # frappe.msgprint(frappe.as_json(delayed_result))
    
    return {
        "completed": project_result[0].completed or 0,
        "in_progress": project_result[0].in_progress or 0,
        "delayed": delayed_result[0].delayed_project or 0
    }

def get_approved_vs_unapproved_projects(project=None):
    query = """
        SELECT 
            COUNT(CASE WHEN approval_status = 'Pending' THEN 1 END) AS pending,
            COUNT(CASE WHEN approval_status = 'Approved By Program Manager' THEN 1 END) AS approved_by_pm,
            COUNT(CASE WHEN approval_status = 'Approved by the CEO' THEN 1 END) AS approved_by_ceo
        FROM `tabProject`
    """
    result = frappe.db.sql(query, as_dict=False)
    
    pending = result[0][0] if result else 0
    approved_by_pm = result[0][1] if result else 0
    approved_by_ceo = result[0][2] if result else 0
    
    return {
        "pending": pending,
        "approved_by_pm": approved_by_pm,
        "approved_by_ceo": approved_by_ceo
    }

def get_total_survey_forms(project=None):
    query = """
        SELECT 
            COUNT(name) as survey_count
        FROM `tabProject`
    """
    result = frappe.db.sql(query, as_dict=False)
    
    return result[0][0] if result else 0

